bin/root-transition.sh

```diff
@@ -0,0 +1,150 @@
+#!/bin/bash
+
+# Copyright 2019 Istio Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script extends the lifetime of the self-signed Citadel root certificate in the current cluster.
+# It offers the following options:
+# - check: Check the current root certificate lifetime.
+# - transition: Extend the lifetime of the current root certificate.
+# - verify: Check the new workload certificates are generated.
+
+trustdomain() {
+  openssl x509 -in ${1} -noout -issuer | cut -f3 -d'='
+}
+
+check_secret () {
+	MD5=`kubectl get secret $1 -o yaml -n $2 | sed -n 's/^.*root-cert.pem: //p' | md5sum | awk '{print $1}'`
+	if [ "$ROOT_CERT_MD5" != "$MD5" ]; then
+		echo "  Secret $2.$1 is NOT updated."
+		NOT_UPDATED="$NOT_UPDATED $2.$1"
+	else
+		echo "  Secret $2.$1 is updated."
+	fi
+}
+
+verify_namespace () {
+	SECRETS=`kubectl get secret -n $1 | grep "istio\.io\/key-and-cert" | awk '{print $1}'`
+	for s in $SECRETS
+	do
+		check_secret $s $1
+	done
+}
+
+verify() {
+  NOT_UPDATED=
+
+  echo "This script checks the current root CA certificate is propagated to all the Istio-managed workload secrets in the cluster."
+
+  ROOT_SECRET=`kubectl get secret istio-ca-secret -o yaml -n istio-system | sed -n 's/^.*ca-cert.pem: //p'`
+  if [ -z $ROOT_SECRET ]; then
+    echo "Root secret is empty. Are you using the self-signed CA?"
+    exit
+  fi
+
+  ROOT_CERT_MD5=`kubectl get secret istio-ca-secret -o yaml -n istio-system | sed -n 's/^.*ca-cert.pem: //p' | md5sum | awk '{print $1}'`
+
+  echo Root cert MD5 is $ROOT_CERT_MD5
+
+  NS=`kubectl get ns | grep -v "STATUS" | grep -v "kube-system" | grep -v "kube-public" | awk '{print $1}'`
+
+  for n in $NS
+  do
+    echo "Checking namespace: $n"
+    verify_namespace $n
+  done
+
+  if [ -z $NOT_UPDATED ]; then
+    echo "------All Istio keys and certificates are updated in secret!"
```

Comment:
If you run verify twice, the message says that `certificate are updated` and that can feel like you just changed something.

It should say "certificates are up to date, or state the expiration date in the future?

Reply:
I ran verify before running anything and it said everything was fine.
Sounds like there is very little difference between check and verify.

Reply:
"Verify" verifies the root certs propagated to the workload secrets are the same as the root cert used in Citadel.
"Check" checks the current root cert lifetime.
I can rename the command if necessary.

Reply:
Agree to improve the message.